### PR TITLE
Bounty System

### PR DESCRIPTION
**Pull request by SciathGH**

### CODE REVIEW

File: Server/src/main/kotlin/net/horizonsend/ion/server/AdminBountyCommands.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import net.horizonsend.ion.common.database.collections.PlayerData
+import net.horizonsend.ion.common.database.update
+import net.horizonsend.ion.server.utilities.feedback.FeedbackType
+import net.horizonsend.ion.server.utilities.feedback.sendFeedbackMessage
+import net.kyori.adventure.text.minimessage.MiniMessage
+import org.bukkit.command.CommandSender
+
+@CommandAlias("adminbounty")
+@Suppress("unused")
+class AdminBountyCommands : BaseCommand() {
```
Comment (Peter-Crawley) [inline Server/src/main/kotlin/net/horizonsend/ion/server/AdminBountyCommands.kt:17]: should be a subcommand of bounty

File: Server/src/main/kotlin/net/horizonsend/ion/server/AdminBountyCommands.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	@Subcommand("add")
+	@CommandPermission("ion.adminbounty.add")
```
Comment (Peter-Crawley) [inline Server/src/main/kotlin/net/horizonsend/ion/server/AdminBountyCommands.kt:19]: `ion.bounty.admin.add`

File: Server/src/main/kotlin/net/horizonsend/ion/server/AdminBountyCommands.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+			return
+		}
+		sender.sendMessage(MiniMessage.miniMessage().deserialize("<gray>Modified $target's bounty by <aqua>$ammount<aqua>"))
+		playerData.update {
+			bounty += ammount
+		}
+	}
+
+	@Subcommand("clear")
+	@CommandPermission("ion.adminbounty.clear")
```
Comment (Peter-Crawley) [inline Server/src/main/kotlin/net/horizonsend/ion/server/AdminBountyCommands.kt:34]: `ion.bounty.admin.clear`

File: Server/src/main/kotlin/net/horizonsend/ion/server/AdminBountyCommands.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		}
+		sender.sendMessage(MiniMessage.miniMessage().deserialize("<gray>Cleared ${target}'s bounty data (accepted bounties and bounty size)"))
+		playerData.update {
+			bounty = 0
+			acceptedBounties = null
+		}
+	}
+
+	@Subcommand("set")
+	@CommandPermission("ion.adminbounty.set")
```
Comment (Peter-Crawley) [inline Server/src/main/kotlin/net/horizonsend/ion/server/AdminBountyCommands.kt:50]: `ion.bounty.admin.set`

File: Server/src/main/kotlin/net/horizonsend/ion/server/IonServer.kt [hunk 5]
```diff
@@ -25,6 +26,8 @@ class IonServer : JavaPlugin() {
 		lateinit var Ion: IonServer private set
 	}
 
+	var bounties: MutableMap<Player, MutableMap<Int, Int>> = mutableMapOf()
```
Comment (Peter-Crawley) [inline Server/src/main/kotlin/net/horizonsend/ion/server/IonServer.kt:29]: `val`

File: Server/src/main/kotlin/net/horizonsend/ion/server/listeners/bukkit/PlayerDeathListener.kt [hunk 6]
```diff
@@ -20,10 +26,11 @@ class PlayerDeathListener : Listener {
 
 	@EventHandler(priority = EventPriority.MONITOR)
 	fun onPlayerDeathEvent(event: PlayerDeathEvent) {
+		event.entity as? ShulkerBullet
```
Comment (Peter-Crawley) [inline Server/src/main/kotlin/net/horizonsend/ion/server/listeners/bukkit/PlayerDeathListener.kt:29]: ???

File: Server/src/main/kotlin/net/horizonsend/ion/server/listeners/bukkit/PlayerDeathListener.kt [hunk 7]
```diff
@@ -40,5 +47,16 @@ class PlayerDeathListener : Listener {
 
 		event.entity.world.dropItem(event.entity.location, head)
 		if (event.player.killer !== event.player) event.player.killer!!.rewardAchievement(Achievement.KILL_PLAYER)
+
+		val killer = event.player.killer
+		val killerData = PlayerData[killer?.uniqueId!!]
```
Comment (Peter-Crawley) [inline Server/src/main/kotlin/net/horizonsend/ion/server/listeners/bukkit/PlayerDeathListener.kt:52]: `killer?.uniqueId ?: return`
avoid an NPE

**Review by Peter-Crawley — CHANGES_REQUESTED** [review]

File: Server/src/main/kotlin/net/horizonsend/ion/server/listeners/bukkit/PlayerDeathListener.kt (hunk 6 above)
Comment (SciathGH) [reply Server/src/main/kotlin/net/horizonsend/ion/server/listeners/bukkit/PlayerDeathListener.kt:29]: dont question it

Comment (Peter-Crawley) [reply Server/src/main/kotlin/net/horizonsend/ion/server/listeners/bukkit/PlayerDeathListener.kt:29]: that isnt a justification

File: Server/src/main/kotlin/net/horizonsend/ion/server/BountyCommands.kt [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	@CommandCompletion("@players")
+	fun onBountyUnset(sender: Player){
+		val playerData = PlayerData[sender.uniqueId]
+		playerData.update {
+			acceptedBounties = null
+		}
+		sender.sendFeedbackMessage(FeedbackType.SUCCESS, "Unset Bounty Successfully")
+	}
+
+	@Subcommand("adminAdd")
```
Comment (Peter-Crawley) [inline Server/src/main/kotlin/net/horizonsend/ion/server/BountyCommands.kt:75]: `admin add`

**Pull request merged**